### PR TITLE
Implement wrapInstance and getCppPointer (#59)

### PR DESCRIPTION
**Pull request by mottosso**
This adds a wrapper for `wrapInstance` and `getCppPointer` from `shiboken2` and automatically unifies the differences with `shiboken` and `sip` for both Python 2 and 3.

| Attribute               | Returns    
|:------------------------|:-----------
| `QtCompat.wrapInstance(addr=long, type=QObject)` | `QObject` |
 `QtCompat.getCppPointer(object=QObject)`    | `long`    

- See #53 for details.
- To **test**, copy/paste [this file](https://raw.githubusercontent.com/abstractfactory/Qt.py/implement59/Qt.py)

**Usage**

```python
import sys
from Qt import QtCompat, QtWidgets
app = QtWidgets.QApplication(sys.argv)
button = QtWidgets.QPushButton("Hello world")
pointer = QtCompat.getCppPointer(button)
widget = QtCompat.wrapInstance(long(pointer), QtWidgets.QWidget)
assert widget == button
app.exit()
```

**Maya Example**

This works for both 2016 and 2017.

```python
import sys
from maya import OpenMayaUI
from Qt import QtCompat, QtWidgets
pointer = OpenMayaUI.MQtUtil.mainWindow()
widget = QtCompat.wrapInstance(long(pointer), QtWidgets.QWidget)
assert isinstance(widget, QtWidgets.QWidget)
```

### Important

This addition requires `sip`, `shiboken` or `shiboken2` to be available on your system. If not found, Qt.py will still import successfully, but these members will not be available.

In such cases, here is a Qt-only version and guaranteed cross-compatible version of the above.

```python
from Qt import QtWidgets
app = QtWidgets.QApplication.instance()
widget = {o.objectName(): o for o in app.topLevelWidgets()}["MayaWindow"]
```

The same pattern may be applied to any and all uses of `sip`, `shiboken` and `shiboken2`, as discussed in-depth at #53.

Enjoy!

### CODE REVIEW

**Comment by mottosso** [comment]
## More Examples

For consideration into the main README.

### Styling the Maya Status Line

**shiboken**

```python
from maya import mel, OpenMayaUI
from Qt import QtWidgets
status_line = mel.eval('$temp=$gStatusLineForm')
ptr = OpenMayaUI.MQtUtil.findControl(status_line)
status_line = shiboken2.wrapInstance(long(ptr), QtWidgets.QWidget)
status_line = status_line.children()[1].children()[1]
status_line.setStyleSheet("QWidget {background: red}")
```

**Qt**

```python
from maya import mel
from Qt import QtWidgets
app = QtWidgets.QApplication.instance()
window = {o.objectName(): o for o in app.topLevelWidgets()}["MayaWindow"]
status_line = mel.eval('$temp=$gStatusLineForm')
status_line = window.findChild(QtGui.QWidget, gStatusLine)
status_lne.setStyleSheet("QWidget {background: red}")
```

### Styling the Channel Box 

**shiboken**

```python
import shiboken
import maya.OpenMayaUI as apiUI
from Qt import QtWidgets
channel_box = apiUI.MQtUtil.findControl("mainChannelBox")
channel_box = shiboken.wrapInstance(long(channel_box), QtWidgets.QTableView)
channel_box.setStyleSheet("QWidget {background: red}")
```

**Qt**

```python
from Qt import QtWidgets
app = QtWidgets.QApplication.instance()
window = {o.objectName(): o for o in app.topLevelWidgets()}["MayaWindow"]
channel_box = window.findChild(QtWidgets.QTableView, "mainChannelBox")
channel_box.setStyleSheet("QWidget {background: green}")
```

**Comment by mottosso** [comment]
What are our thoughts on this?

- http://nathanhorne.com/pyqtpyside-wrap-instance

He isn't pointing out what the difference is exactly, anyone experienced this problem?

**Comment by mottosso** [comment]
> He isn't pointing out what the difference is exactly, anyone experienced this problem?

His code isn't true. `sip` **does not** support what he's implemented for shiboken.

```python
>>> import sip
>>> import sys
>>> from PyQt4 import QtGui
>>> app = QtGui.QApplication(sys.argv)
>>> button = QtGui.QPushButton("Hello world")
>>> pointer = sip.unwrapinstance(button)
>>> sip.wrapinstance(long(pointer))
# Traceback (most recent call last):
#   File "<stdin>", line 1, in <module>
# TypeError: wrapinstance() takes exactly 2 arguments (1 given)
>>> sip.wrapinstance(long(pointer), None)
# Traceback (most recent call last):
#   File "<stdin>", line 1, in <module>
# TypeError: must be sip.wrappertype, not None
```

I've implemented this regardless, as I can't see why not and it makes for less code. The feature is ignored when used *with* a `base` argument and kicks in automatically when passing it.

```python
import sys
from Qt import QtCompat, QtWidgets, __binding__

app = QtWidgets.QApplication(sys.argv)

try:
    button = QtWidgets.QPushButton("Hello world")
    button.setObjectName("MySpecialButton")
    pointer = QtCompat.getCppPointer(button)
    widget = QtCompat.wrapInstance(long(pointer))
    assert isinstance(widget, QtWidgets.QWidget), widget
    assert widget.objectName() == button.objectName()

    if __binding__ in ("PySide", "PySide2"):
        assert widget != button
    else:
        assert widget == button
finally:
    app.exit()
```

In either case, I uncovered a subtle difference between `sip` and `shiboken` whilst writing the tests. 
Notice, the objects are not directly comparable in PySide 1 and 2, but are in PyQt4 and PyQt5.

**Comment by dgovil** [comment]
Hi @mottosso I just wanted to check what the current state of this PR was in your view? What still needs doing?

I was considering merging this into our local copy of `Qt.py` till the PR was merged unless you think there will be major changes?

**Comment by mottosso** [comment]
Hey @dgovil, no major changes as far as I can see, what's missing now is testing.

> I was considering merging this into our local copy of Qt.py

Please do, after some use, if you haven't encountered any issues, that'd be a good point to merge this.

File: Qt.py [hunk 1]
```diff
@@ -645,6 +645,50 @@ def _setup(module, extras):
...
+
+    Arguments:
+        func (function): Original function
+        ptr (long): Pointer to QObject in memory
+        base (QObject, optional): Base class to wrap with. Defaults to QObject,
+            which should handle anything.
+
+    """
+
+    assert isinstance(ptr, long), "Argument 'ptr' must be of type <long>"
```
Comment (dgovil) [inline Qt.py]: As far as I know, in Python3 there is no longer a `long`, and instead it should be an int in python3.

Comment (mottosso) [inline Qt.py]: Hm, you're right. This wasn't caught by Travis, I'll take a closer look as to why that is tomorrow.

**Comment by mottosso** [comment]
### Get top-level window in any binding and any application.

sip and shiboken is sometimes used to fetch the main window of an application in order to make it a parent of a custom window. Below is an example of how to find said window efficiently and in any situation.

```python
from Qt import QtWidgets

current = QtWidgets.QApplication.activeWindow()
while current:
    parent = current
    current = parent.parent()

print(parent)
```

File: Qt.py (hunk 1 above)
Comment (mottosso) [inline Qt.py]: Ok, this has been fixed.

Turns out `shiboken` isn't supported by Python 3.5.
https://github.com/PySide/shiboken-setup/issues/3

**Comment by dgovil** [comment]
FWIW, been running this in production for a week and it seems to be working correctly after porting some existing tools over to using this.

**Comment by mottosso** [comment]
That's great @dgovil, thanks for reporting back.

File: Qt.py [hunk 2]
```diff
@@ -625,6 +629,50 @@ def _setup(module, extras):
...
+        2. A `base` argument is not provided.
+
+        See :func:`QtCompat.wrapInstance()`
+
+    Arguments:
+        func (function): Original function
+        ptr (long): Pointer to QObject in memory
+        base (QObject, optional): Base class to wrap with. Defaults to QObject,
+            which should handle anything.
+
```
Comment (fredrikaverpil) [inline Qt.py]: Perhaps we should add a "Returns:" to the docstring to make it easier to understand the different types of return values are possible here?

File: Qt.py [hunk 3]
```diff
@@ -625,6 +629,50 @@ def _setup(module, extras):
...
+        super_class_name = meta_object.superClass().className()
+
+        if hasattr(Qt.QtWidgets, class_name):
+            base = getattr(Qt.QtWidgets, class_name)
+
+        elif hasattr(Qt.QtWidgets, super_class_name):
+            base = getattr(Qt.QtWidgets, super_class_name)
+
+        else:
+            base = Qt.QtCore.QObject
```
Comment (fredrikaverpil) [inline Qt.py]: Out of curiosity, in which scenario(s) is the base set to `Qt.QtCore.QObject`?
Is it for PyQt4 and PySide?

**Comment by fredrikaverpil** [comment]
I'm sorry I've been so absent in regards to this addition to Qt.py.

This is a fantastic PR. Crossing my fingers we won't have to spend too much time maintaining this though. :wink:

Before merging, I'd like to test this Qt.py version internally. We shoud also revise the [`loadUi` example](https://github.com/abstractfactory/Qt.py/tree/master/examples/load_ui).

Really nice job, @mottosso and @dgovil!

File: Qt.py (hunk 3 above)
Comment (mottosso) [inline Qt.py]: Likely never, all objects you could get from this function are subclasses of QObject. It's set as default here as a fallback, as anything, even a `QWidget` or your own subclass of a subclass of `QMainWindow` could be interpreted as `QObject`. In those cases, you still gain access to various superclass properties like `objectName()` but not the more specialised methods like `resize()` etc.

File: Qt.py (hunk 2 above)
Comment (mottosso) [inline Qt.py]: Yes, and `long` must also change.

Not sure how to tackle the fact that `long` is absent from Python 3. I had a look at us internally taking `int` for both Python 2 and 3, but then we stray from the original `PySide2.wrapInstance` argument signature.

**Comment by fredrikaverpil** [comment]
I'll look into this but I'm getting an error in Nuke:

```python
class MyWindow(QtWidgets.QMainWindow):
    def __init__(self, parent=None, resize=[]):
        super(MyWindow, self).__init__(parent)
        main_window_file = os.path.join(ui_dir, 'main_window.ui')
        self.main_widget = QtCompat.loadUi(main_window_file)  # errors as seen below
```

```python
Traceback (most recent call last):
  File "C:/Users/iruser/skalman/cache/1.2.3.4/pipeline/bin/nuke/builds/nuke_v9.0v9_win64/plugins\nukescripts\panels.py", line 153, in makeUI
    self.widget = self.widgetClass()
  File "c:\users\iruser\code\repos\skalman\skalman\uisetup.py", line 154, in __init__
    self.main_widget = QtCompat.loadUi(main_window_file)
  File "C:\Users\iruser\skalman\condaenvs\skalman_py27\Lib\site-packages\Qt.py", line 1028, in _loadUi
    Qt.QtCore.QMetaObject.connectSlotsByName(widget)
RuntimeError: Internal C++ object (PySide.QtGui.QMainWindow) already deleted.
```

**Comment by fredrikaverpil** [comment]
Ok, here's a reproducible.

This runs just fine in 1.0.0b3 but errors using Qt.py from this PR:

```python
import tempfile
import os
import io

import Qt
from Qt import QtWidgets
from Qt import QtCompat

import nuke
import nukescripts


TEMPDIR = tempfile.mkdtemp()
tempfile = os.path.join(TEMPDIR, "qmainwindow.ui")

with io.open(tempfile, "w", encoding="utf-8") as f:
    f.write(u"""\
<?xml version="1.0" encoding="UTF-8"?>
<ui version="4.0">
 <class>MainWindow</class>
 <widget class="QMainWindow" name="MainWindow">
  <property name="geometry">
   <rect>
    <x>0</x>
    <y>0</y>
    <width>223</width>
    <height>140</height>
   </rect>
  </property>
  <property name="windowTitle">
   <string>MainWindow</string>
  </property>
  <widget class="QWidget" name="centralwidget">
   <widget class="QPushButton" name="pushButton">
    <property name="geometry">
     <rect>
      <x>70</x>
      <y>40</y>
      <width>80</width>
      <height>16</height>
     </rect>
    </property>
    <property name="text">
     <string>PushButton</string>
    </property>
   </widget>
  </widget>
  <widget class="QMenuBar" name="menubar">
   <property name="geometry">
    <rect>
     <x>0</x>
     <y>0</y>
     <width>223</width>
     <height>17</height>
    </rect>
   </property>
  </widget>
  <widget class="QStatusBar" name="statusbar"/>
 </widget>
 <resources/>
 <connections/>
</ui>
""")


class MyWindow(QtWidgets.QMainWindow):
    def __init__(self, parent=None, resize=[]):
        super(MyWindow, self).__init__(parent)
        self.main_widget = QtCompat.loadUi(tempfile)
        self.setCentralWidget(self.main_widget)

# assert hasattr(Qt, '_wrapinstance')  # Make sure we are using abstractfactory:implement59

panel = nukescripts.panels.registerWidgetAsPanel(
    widget='MyWindow',
    name='SomeWindow',
    id='uk.co.thefoundry.' + 'SomeWindow',
    create=True)

pane = nuke.getPaneFor('Properties.1')
panel.addToPane(pane)
my_window_widget = panel.customKnob.getObject().widget

assert my_window_widget
```

**Comment by fredrikaverpil** [comment]
Aha, this was introduced in 1.0.0b4 and not in this PR. I'll file a separate issue.

**Comment by fredrikaverpil** [comment]
Here's a slightly more portable piece of reproducible code which doesn't require Nuke:

```python
import tempfile
import os
import io

from Qt import QtWidgets
from Qt import QtCompat

TEMPDIR = tempfile.mkdtemp()
tempfile = os.path.join(TEMPDIR, "qmainwindow.ui")

with io.open(tempfile, "w", encoding="utf-8") as f:
    f.write(u"""\
<?xml version="1.0" encoding="UTF-8"?>
<ui version="4.0">
 <class>MainWindow</class>
 <widget class="QMainWindow" name="MainWindow">
  <property name="geometry">
   <rect>
    <x>0</x>
    <y>0</y>
    <width>223</width>
    <height>140</height>
   </rect>
  </property>
  <property name="windowTitle">
   <string>MainWindow</string>
  </property>
  <widget class="QWidget" name="centralwidget">
   <widget class="QPushButton" name="pushButton">
    <property name="geometry">
     <rect>
      <x>70</x>
      <y>40</y>
      <width>80</width>
      <height>16</height>
     </rect>
    </property>
    <property name="text">
     <string>PushButton</string>
    </property>
   </widget>
  </widget>
  <widget class="QMenuBar" name="menubar">
   <property name="geometry">
    <rect>
     <x>0</x>
     <y>0</y>
     <width>223</width>
     <height>17</height>
    </rect>
   </property>
  </widget>
  <widget class="QStatusBar" name="statusbar"/>
 </widget>
 <resources/>
 <connections/>
</ui>
""")


class MyWindow(QtWidgets.QMainWindow):
    def __init__(self, parent=None):
        super(MyWindow, self).__init__(parent)
        self.main_widget = QtCompat.loadUi(tempfile)
        self.setCentralWidget(self.main_widget)

my_window = MyWindow()
my_window.show()
```

**Comment by dgovil** [comment]
I can't replicate that error with the PySide2 version for Maya 2017, or PySide 1.2.2 (Qt 4.8.5) and  or PyQt4.10.1 (Qt 4.8.5). Haven't tried with PyQt5 yet though.
They all show me a window with a PushButton, and the other ui elements are available.

Granted I'm trying your last example standalone without Nuke being involved.

**Comment by fredrikaverpil** [comment]
@dgovil The code snippet I gave works fine everywhere but in Nuke. Since this was introduced in a previous version (and not in this PR) I'll file a separate issue.

**Comment by dgovil** [comment]
@mottosso I was wondering where wrapInstance ended up into making it into `Qt.py`? We've been using it in production for a while now without issue so I was wondering if it is possible to get it into `Qt.py`. Right now I am managing our own fork but I'd love to get it down to as few changes as possible

**Comment by mottosso** [comment]
> We've been using it in production for a while now without issue so I was wondering if it is possible to get it into Qt.py.

That's good enough for me.

If you could have a look at merging this with the current state of the repo then we should be able to have it merged by dinner time.

**Pull request merged**